client/history/redirector.go

```diff
@@ -0,0 +1,146 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"context"
+	"fmt"
+
+	"go.temporal.io/api/serviceerror"
+
+	"go.temporal.io/server/api/historyservice/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/convert"
+	"go.temporal.io/server/common/membership"
+	serviceerrors "go.temporal.io/server/common/serviceerror"
+)
+
+type (
+	// An operationTarget represents the intended destination for a history
+	// client operation. If shardID == 0, the operation is for a specific history
+	// instance at address. Otherwise, the operation is intended for whichever
+	// instance owns the shard.
+	operationTarget struct {
+		shardID int32
+		address rpcAddress
+	}
+
+	// A redirector executes a client operation against a history instance.
+	// If the operation is intended for the owner of a shard, and the request
+	// returns a shard ownership lost error with a hint for a new shard owner,
+	// the redirector will retry the request to the new owner.
+	redirector interface {
```

Comment:
I'd remove this interface and just rely on the struct. This is just an internal helper, so it doesn't make sense for it to be pluggable, and we shouldn't ever mock it because it's just our internal logic--not some I/O interface.

Reply:
The next PR I have staged will make the implementation for this configurable, as that's where I'm putting in the new shard ownership handling logic. I do want to make that configurable so it can enabled or disabled.